src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorCodeDocument.cs

```diff
@@ -17,41 +17,108 @@ public sealed partial class RazorCodeDocument
+    private readonly RazorSyntaxTree? _syntaxTree;
+    private readonly ImmutableArray<RazorSyntaxTree> _importSyntaxTrees;
+    private readonly TagHelperDocumentContext? _tagHelperContext;
+    private readonly DocumentIntermediateNode? _documentNode;
+    private readonly RazorCSharpDocument? _cSharpDocument;
+
+    private readonly Lazy<RazorHtmlDocument> _htmlDocument;
 
     private RazorCodeDocument(
         RazorSourceDocument source,
         ImmutableArray<RazorSourceDocument> imports,
-        RazorParserOptions? parserOptions,
-        RazorCodeGenerationOptions? codeGenerationOptions,
-        PropertyTable? properties = null)
+        RazorParserOptions parserOptions,
+        RazorCodeGenerationOptions codeGenerationOptions,
+        TagHelperCollection? tagHelpers,
+        TagHelperCollection? referencedTagHelpers,
+        RazorSyntaxTree? preTagHelperSyntaxTree,
+        RazorSyntaxTree? syntaxTree,
+        ImmutableArray<RazorSyntaxTree> importSyntaxTrees,
+        TagHelperDocumentContext? tagHelperContext,
+        DocumentIntermediateNode? documentNode,
+        RazorCSharpDocument? cSharpDocument)
     {
         Source = source;
         Imports = imports.NullToEmpty();
 
-        ParserOptions = parserOptions ?? RazorParserOptions.Default;
-        CodeGenerationOptions = codeGenerationOptions ?? RazorCodeGenerationOptions.Default;
+        ParserOptions = parserOptions;
+        CodeGenerationOptions = codeGenerationOptions;
 
-        _properties = properties ?? new();
+        _tagHelpers = tagHelpers;
+        _referencedTagHelpers = referencedTagHelpers;
+        _preTagHelperSyntaxTree = preTagHelperSyntaxTree;
+        _syntaxTree = syntaxTree;
+        _importSyntaxTrees = importSyntaxTrees;
+        _tagHelperContext = tagHelperContext;
+        _documentNode = documentNode;
+        _cSharpDocument = cSharpDocument;
+
+        _htmlDocument = new Lazy<RazorHtmlDocument>(() => RazorHtmlWriter.GetHtmlDocument(this));
     }
 
-    public static RazorCodeDocument Create(
-        RazorSourceDocument source,
+    private RazorCodeDocument With(
+        RazorSourceDocument? source = null,
+        ImmutableArray<RazorSourceDocument>? imports = null,
         RazorParserOptions? parserOptions = null,
-        RazorCodeGenerationOptions? codeGenerationOptions = null)
+        RazorCodeGenerationOptions? codeGenerationOptions = null,
+        TagHelperCollection? tagHelpers = null,
+        TagHelperCollection? referencedTagHelpers = null,
+        RazorSyntaxTree? preTagHelperSyntaxTree = null,
+        RazorSyntaxTree? syntaxTree = null,
+        ImmutableArray<RazorSyntaxTree>? importSyntaxTrees = null,
+        TagHelperDocumentContext? tagHelperContext = null,
+        DocumentIntermediateNode? documentNode = null,
+        RazorCSharpDocument? cSharpDocument = null)
     {
-        ArgHelper.ThrowIfNull(source);
+        if (isChanged(source, Source)
+            || (imports is not null)
```

Comment:
>(imports is not null) [](http://example.com/codeflow?start=15&length=21)

Could you change the ImmutableArray checks to check the backing arrays? Something like:

```
        || isArrayChanged(imports, Imports)

...

        static bool isArrayChanged<T>(ImmutableArray<T>? newArray, ImmutableArray<T> oldArray)
        {
            return newArray is null
                ? oldArray.Length > 0
                : !object.ReferenceEquals(ImmutableCollectionsMarshal.AsArray<T>(newArray.Value), ImmutableCollectionsMarshal.AsArray<T>(oldArray));
        }
```

Reply:
I don't get the local functions for comparisons anyway. Shouldn't we just be implementing IEquatable on all of these subtypes, and using Equals and SequenceEquals or something?